File: 20.有效的括号.py
#
# @lc app=leetcode.cn id=20 lang=python3
#
# [20] 有效的括号
#

# @lc code=start


class Solution:
    def isValid(self, s: str) -> bool:
        dic = {"{": "}", "[": "]", "(": ")", "?": "?"}
        stack = ["?"]  # stack为空时pop会报错, 所以一开始添加一个 ? 元素
        for c in s:
            if c in dic:
                stack.append(c)
            elif dic[stack.pop()] != c:
                return False
        return len(stack) == 1


# s = Solution()
# r = s.isValid("]")
# print(r)

class Solution:
    def isValid(self, s: str) -> bool:
        dic = {"}": "{", "]": "[", ")": "("}
        stack = []
        for char in s:
            if char in dic.values():
                stack.append(char)
            elif char in dic.keys():
                if stack == [] or dic[char] != stack.pop():
                    return False
            else:
                return False
        return stack == []


class Solution:
    def isValid(self, s: str) -> bool:
        while '{}' in s or '()' in s or '[]' in s:
            s = s.replace('{}', '')
            s = s.replace('[]', '')
            s = s.replace('()', '')
        return s == ''


# @lc code=end

File: 20.有效的括号.js
/*
 * @lc app=leetcode.cn id=20 lang=javascript
 *
 * [20] 有效的括号
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  let dict = { '{': '}', '(': ')', '[': ']' }
  let stack = []
  for (const c of s) {
    // 左括号 压栈
    if (c in dict) {
      stack.push(c)

    }
    // 右括号 出栈, 判断和栈顶元素是不是匹配的一对括号
    else if (dict[stack.pop()] != c) {
      return false
    }
  }
  return stack.length === 0
};

// @lc code=end

